src/ophyd_async/core/utils.py

```diff
@@ -11,46 +24,82 @@
 #: monitor updates
 ReadingValueCallback = Callable[[Reading, T], None]
 DEFAULT_TIMEOUT = 10.0
+ErrorText = Union[str, Dict[str, Exception]]
 
 
 class NotConnected(Exception):
     """Exception to be raised if a `Device.connect` is cancelled"""
 
-    def __init__(self, *lines: str):
-        self.lines = list(lines)
+    _indent_width = "    "
+
+    def __init__(self, errors: ErrorText):
+        """
+        NotConnected holds a mapping of device/signal names to
+        errors.
+
+        Parameters
+        ----------
+        errors: ErrorText
+            Mapping of device name to Exception or another NotConnected.
+            Alternatively a string with the signal error text.
+        """
+
+        self._errors = errors
+
+    def _format_exception_string(self, exception: Exception):
+        exception_name = f" {type(exception).__name__}"
+        exception_text = f": {exc_text}\n" if (exc_text := str(exception)) else "\n"
+        return exception_name + exception_text
+
+    def _format_error_string(self, errors: ErrorText, indent="") -> str:
+        string = ""
+        if not isinstance(errors, dict) and not isinstance(errors, str):
+            raise RuntimeError(
+                f"Unknown error type `{type(errors)}` " "expected `str` or `dict`"
+            )
+        if isinstance(errors, str):
+            return errors
+
+        for name, value in errors.items():
+            string += indent + f"{name}:"
+            if isinstance(value, NotConnected):
+                if isinstance(value._errors, str):
+                    string += self._format_exception_string(value)
+                else:
+                    string += "\n" + value._format_error_string(
+                        value._errors, indent=(indent + self._indent_width)
+                    )
+            elif isinstance(value, Exception):
+                string += self._format_exception_string(value)
+            else:
+                raise RuntimeError(f"`{type(value)}` not a string or exception")
```

Comment:
If we allow formatting like:
```
device: NotConnected:
    sigA: NotConnected: ca://...
    sigB: ValueError: 32
```

Then we can simplify to

```suggestion
        for name, value in errors.items():
            if isinstance(value, NotConnected):
                error_txt = value.format_error_string(indent + self._indent_width)
            else:
                error_txt = str(error)
            string += f"\n{indent}{name}: {type(error).__name__}"
            if error_txt:
                string += f": {error_txt}"
```